docs/architecture/adr-012-peer-transport.md

```diff
@@ -0,0 +1,108 @@
+# ADR 012: PeerTransport
+
+## Context
+
+One of the more apparent problems with the current architecture in the p2p
+package is that there is no clear separation of concerns between different
+components. Most notably the `Switch` is currently doing physical connection
+handling. An artifact is the dependency of the Switch on `config.P2PConfig`.
+
+Addresses:
+* [#2046](https://github.com/tendermint/tendermint/issues/2046)
+* [#2047](https://github.com/tendermint/tendermint/issues/2047)
+
+First iteraton in [#2067](https://github.com/tendermint/tendermint/issues/2067)
+
+## Decision
+
+Transport concerns will be handled by a new component (`PeerTransport`) which
+will provide Peers at its boundary to the caller. In turn `Switch` will use
+this new component accept new `Peer`s and dial them based on `NetAddress`.
+
+### PeerTransport
+
+Responsible for emitting and connecting to Peers. The implementation of `Peer`
+is left to the transport, which implies that the chosen transport dictates the
+characteristics of the implementation handed back to the `Switch`. It is the
+place where we enforce low-level guards[0]:
+
+* connections from our own node
+* handshake fails
+* upgrade to secret connection fails
+* protocol/app/... version doesn't match
+
+
+``` go
+// PeerTransport proxies incoming and outgoing peer connections.
+type PeerTransport interface {
+	// Accept returns a newly connected Peer.
+	Accept() (Peer, error)
```

Comment:
I'm having trouble understanding `Accept`. Is `Accept` supposed to evaluate an incoming connection? If so why doesn't have connection as a parameter

Reply:
> Accept is the way for the caller to signal that they are ready to deal with one more incoming Peer.

Ohh I see. Perhaps this may be more clearly named as "GetPeer"?

Reply:
Trying to avoid getter semantics when possible, in the domain of transport accept and dial are very well understood and the signature already communicates what is returned.

Reply:
Sounds good! Thank you for the explanations :)